sklearn/feature_extraction/text.py

```diff
@@ -1117,21 +1117,30 @@ def fit(self, X, y=None):
         X : sparse matrix, [n_samples, n_features]
             a matrix of term/token counts
         """
+        X = check_array(X, accept_sparse='csc')
+
         if not sp.issparse(X):
             X = sp.csc_matrix(X)
+
         if self.use_idf:
             n_samples, n_features = X.shape
-            df = _document_frequency(X)
+            df = _document_frequency(X).astype(X.dtype
+                                               if X.dtype in FLOAT_DTYPES
+                                               else np.float64)
 
             # perform idf smoothing if required
             df += int(self.smooth_idf)
             n_samples += int(self.smooth_idf)
 
             # log+1 instead of log makes sure terms with zero idf don't get
             # suppressed entirely.
-            idf = np.log(float(n_samples) / df) + 1.0
-            self._idf_diag = sp.spdiags(idf, diags=0, m=n_features,
-                                        n=n_features, format='csr')
+            idf = np.log(n_samples / df) + 1
+            self._idf_diag = sp.diags(idf, offsets=0,
+                                      shape=(n_features, n_features),
+                                      format='csr',
+                                      dtype=X.dtype
+                                      if X.dtype in FLOAT_DTYPES
+                                      else np.float64)
```

Comment:
Maybe factorize this as `df_dtype = X.dtype if ...` as it's used twice and is a bit verbose